compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/GroupByKeyAndWindowDoFnTransform.java

```diff
@@ -314,4 +345,31 @@ public TimerInternals timerInternalsForKey(final K key) {
       return stateAndTimerForKey.timerInternals;
     }
   }
+
+  /**
+   * This class wraps the output collector to track the watermark hold of each key.
+   */
+  final class GBKWOutputCollector implements OutputCollector<WindowedValue<KV<K, Iterable<InputT>>>> {
+    private final OutputCollector<WindowedValue<KV<K, Iterable<InputT>>>> outputCollector;
+    GBKWOutputCollector(final OutputCollector<WindowedValue<KV<K, Iterable<InputT>>>> outputCollector) {
+      this.outputCollector = outputCollector;
+    }
+
+    @Override
+    public void emit(final WindowedValue<KV<K, Iterable<InputT>>> output) {
+      // adds the output timestamp to the watermark hold of each key
+      // +1 to the output timestamp because if the window is [0-5000), the timestamp is 4999
+      keyAndWatermarkHoldMap.put(output.getValue().getKey(),
+        new Watermark(output.getTimestamp().getMillis() + 1));
```

Comment:
Is there a reference to this? (e.g., another Beam runner doing also the same)

In the case of following sequence of elements/watermark with (timestamp):

e1(1), e2(2), w1(3), e3(3)

With the +1 we see it as

e1(2=1+1), e2(3=2+1), w1(3), e3(4=3+1)

Would that result in an incorrect reordering of e2 and w1 (in the time-scale)?

Reply:
The output here is  windowed values, not all elements.

Reply:
Can you add a TODO here to check for different triggers?
https://issues.apache.org/jira/browse/NEMO-270

In the following case
- W-4 / E-3 / E-2 / E-1 / E-1 --> GBKW
- All elements are in the window of [1~4)

(1) If triggered at watermark W-4 this works perfectly
(2) If triggered early at E-1, then the downstream operators will consider the following E-1 late.

NEMO-270 can check for the correctness of (2)